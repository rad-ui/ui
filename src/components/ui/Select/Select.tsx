'use client';
import React from 'react';
import SelectContent from './fragments/SelectContent';
import SelectItem from './fragments/SelectItem';
import SelectTrigger from './fragments/SelectTrigger';
import SelectRoot from './fragments/SelectRoot';
import SelectIndicator from './fragments/SelectIndicator';
import SelectPortal from './fragments/SelectPortal';
import SelectGroup from './fragments/SelectGroup';
import SelectSearch from './fragments/SelectSearch';

type SelectComponent = React.ForwardRefExoticComponent<React.ComponentPropsWithoutRef<'div'> & React.RefAttributes<HTMLDivElement>> & {
    Root: typeof SelectRoot;
    Content: typeof SelectContent;
    Item: typeof SelectItem;
    Trigger: typeof SelectTrigger;
    Portal: typeof SelectPortal;
    Group: typeof SelectGroup;
    Indicator: typeof SelectIndicator;
    Search: typeof SelectSearch;
};

const Select = React.forwardRef<HTMLDivElement, React.ComponentPropsWithoutRef<'div'>>((_, ref) => {
    console.warn('Direct usage of Select is not supported. Please use Select.Root, Select.Content, etc. instead.');
    return null;
}) as SelectComponent;
Select.displayName = 'Select';

Select.Root = SelectRoot;
Select.Content = SelectContent;
Select.Item = SelectItem;
Select.Trigger = SelectTrigger;
Select.Portal = SelectPortal;
Select.Group = SelectGroup;
Select.Indicator = SelectIndicator;
Select.Search = SelectSearch;

export default Select;
