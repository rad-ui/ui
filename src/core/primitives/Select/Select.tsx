import SelectPrimitiveContent from './fragments/SelectPrimitiveContent';
import SelectPrimitiveItem from './fragments/SelectPrimitiveItem';
import SelectPrimitiveTrigger from './fragments/SelectPrimitiveTrigger';
import SelectPrimitiveRoot from './fragments/SelectPrimitiveRoot';
import SelectPrimitivePortal from './fragments/SelectPrimitivePortal';
import SelectPrimitiveOverlay from './fragments/SelectPrimitiveOverlay';

const SelectPrimitive = () => {
    console.warn('Direct usage of Select is not supported. Please use Select.Root, Select.Content, etc. instead.');
    return null;
};

SelectPrimitive.Root = SelectPrimitiveRoot;
SelectPrimitive.Content = SelectPrimitiveContent;
SelectPrimitive.Portal = SelectPrimitivePortal;
SelectPrimitive.Overlay = SelectPrimitiveOverlay;
SelectPrimitive.Item = SelectPrimitiveItem;
SelectPrimitive.Trigger = SelectPrimitiveTrigger;

export default SelectPrimitive;
