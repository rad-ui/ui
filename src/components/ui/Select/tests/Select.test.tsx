import React from 'react';
import { render } from '@testing-library/react';
import Select from '../Select';

describe('Select Component', () => {
    test('forwards ref in Select.Root', () => {
        const ref = React.createRef<HTMLDivElement>();
        render(
            <Select.Root ref={ref}>
                <Select.Trigger>Trigger</Select.Trigger>
            </Select.Root>
        );
        expect(ref.current).toBeInstanceOf(HTMLDivElement);
    });

    test('forwards ref in Select.Trigger', () => {
        const ref = React.createRef<HTMLButtonElement>();
        render(
            <Select.Root>
                <Select.Trigger ref={ref}>Trigger</Select.Trigger>
            </Select.Root>
        );
        expect(ref.current).toBeInstanceOf(HTMLButtonElement);
    });

    test('renders hidden native select for accessibility inside forms', () => {
        const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
        const { container } = render(
            <form>
                <Select.Root name="test">
                    <Select.Trigger>Trigger</Select.Trigger>
                    <Select.Content>
                        <Select.Item value="one">One</Select.Item>
                    </Select.Content>
                </Select.Root>
            </form>
        );
        const hiddenSelect = container.querySelector('select[aria-hidden="true"]');
        expect(hiddenSelect).toBeInTheDocument();
        errorSpy.mockRestore();
    });

    test('renders without warnings', () => {
        const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});
        const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
        render(
            <Select.Root>
                <Select.Trigger>Trigger</Select.Trigger>
            </Select.Root>
        );
        expect(warnSpy).not.toHaveBeenCalled();
        expect(errorSpy).not.toHaveBeenCalled();
        warnSpy.mockRestore();
        errorSpy.mockRestore();
    });

    test('matches snapshot', () => {
        const { asFragment } = render(
            <Select.Root>
                <Select.Trigger>Trigger</Select.Trigger>
            </Select.Root>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
