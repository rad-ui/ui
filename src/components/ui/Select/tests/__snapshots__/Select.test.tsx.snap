// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rad-ui-select-root"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="rad-ui-select-trigger"
      role="combobox"
      type="button"
    >
      Trigger
    </button>
  </div>
</DocumentFragment>
`;
